### ISSUE TITLE
Allow adding and editing of more complex OSM types

### ISSUE THREAD

**Issue by biodranik** [organicmaps/organicmaps#4523]
Now we map OSM types to ours using data/mapcss-mapping.csv, for example:
```
man_made|tower|communication;[man_made=tower][tower:type=communication];;name;int_name;311;
```
Here our type `man_made-tower-communication` is mapped to a pair of OSM tags `man_made=tower` and `tower:type=communication`. The Editor doesn't know about it.

Some original OSM types are not automatically "guessed" by the Editor upload code. To properly support these cases, the following idea can be implemented:

Extend our existing data/editor.config XML format to contain the appropriate mapping of our types back to OSM types, so the Editor can correctly check if the corresponding feature already exists in OSM, match it, and update or create properly.

**Comment by biodranik** [comment]
A user requested to add mosques using the Editor: amenity=place_of_worship religion=muslim

**Comment by pastk** [comment]
ATM we have a special hardcoded support for `amenity-recycling-centre => amenity=recycling recycling_type=centre` and `amenity-recycling-container => amenity=recycling recycling_type=container`, see in `editor/xml_feature.cpp::ToXML()` and `::FromXML()`.

**Comment by pastk** [comment]
> Extend our existing data/editor.config XML format to contain the appropriate mapping of our types back to OSM types, so the Editor can correctly check if the corresponding feature already exists in OSM, match it, and update or create properly.

I wonder if the Editor correctly handles cases of **simple** (one tag) types, but which can have several OSM types options, e.g.
`shop|clothes;[shop=clothes],[shop=fashion]` or when `replaced_tags.txt` is used?

FYI @vng

**Comment by pastk** [comment]
> 
> Extend our existing data/editor.config XML format to contain the appropriate mapping of our types back to OSM types, so the Editor can correctly check if the corresponding feature already exists in OSM, match it, and update or create properly.

It would lead to duplication of matching info in the editor config and mapping-mapcss, two sources of truth basically. Its more work and inevitable sync issues.

Instead the Editor should take the right matching directly from mapping-mapcss or a simpler/smaller file automatically derived from it. This way it'll be possible to support multiple matching options also, e.g. `shop|clothes;[shop=clothes],[shop=fashion]` (only the first "main" option will be used for adding new features, other options might be used for editing existing features).

**Comment by hemanggs** [comment]
The bottleneck  in the editor from my understanding

- OM types that have multiple primary OSM tags can be added using mapcss_mapping.csv & editor.config , the problem is in the upload part of the code . The current XML conversion logic (in xml_feature.cpp , `ToXML` `TypeToXML` `FromXML`)   gets the OSM tags by either string splitting (works for trivial OM types to OSM tags mapping eg : `amenity-restaurant` -> `amenity=restaurant` ) or complex tags have to be hardcoded which the crux of the issue. (eg : `amenity-recycling-centre` & `man_made-tower-communication`

Idea for the fix : 

- We can use mapcss-mapping.csv as single source of truth and create a derived JSON file from it that contains OM types to OSM tags mapping, it gets placed into assets folder during the build process using a python script ,and have a `editor::OsmMapper` class in the core that is responsible for loading this JSON file at runtime ? 

Is this the complete scope for this issue ?

CC @biodranik

**Comment by biodranik** [comment]
Likely yes. Moments to think of:
- not all types are editable 
- is keeping all relationships in one place like edits.xml simpler/better? Note that implementation allows to download edits.xml with emergency fixes to disable some editor features in case of vandalism. 
- merger part requires back and forth type conversions, uploader needs to know if it is a conflict, and be able to resolve it or notify about it

**Comment by hemanggs** [comment]
> * not all types are editable
> * is keeping all relationships in one place like edits.xml simpler/better?

Yeah this could be explored, instead of having a separate JSON combining it with existing edits.xml could be simpler.

>  Note that implementation allows to download edits.xml with emergency fixes to disable some editor features in case of vandalism.

yeah thanks for pointing this out , Its important to incorporate checks specially in the `FromXML()` part, as you mentioned edits.xml is accessible to anyone and without the checks injecting unwanted XML into the editor is fairly easy .

> * merger part requires back and forth type conversions, uploader needs to know if it is a conflict, and be able to resolve it or notify about it

do you mean the changeset_wrapper part of the upload code that does back and forth with the OSM API to check for conflicts before upload ?

**Comment by biodranik** [comment]
Now om types are converted to xml tags, and then xml tags are merged with xml tags from OpenStreetMap.org. 

The tricky part for multi tag types is to understand if the type in OpenStreetMap.org has changed or not. If the type in OpenStreetMap.org changed, but in OM only some data is updated, then likely om changes should be ignored.
